RFC: Login UX

wchristian opened the pull request:
As promised in my email, some attempts at improving the code. I don't expect you to pull this in as is, but think you may find some value in reading it.

I imported the whole thing in Android Studio, did some basic house cleaning, then went to improve the user experience of the login process to look less sketchy:

- always show a detail error, even if it's just a stringified exception
- catch and display common errors

I would've liked to dig a little deeper, but as noted in #23, documentation on how to actually get the app to talk properly to the api is a little lacking.

wchristian committed:
add generated files to .gitignore

wchristian committed:
update lib versions to latest

wchristian committed:
show an error if client secret/id are missing

wchristian committed:
typos

wchristian committed:
clean imports

wchristian committed:
deduplicate error methods

charlag commented on app/src/main/java/eu/theinvaded/mastondroid/viewmodel/LoginViewModel.java:
```diff
@@ -82,41 +82,58 @@ public void destroy() {
+        }
+    }
+
+    private String ensureNonNullString(String string, String error) throws MastodroidLoginSetupException {
```
I think this may go to some StringUtils class instead. Looks generic enough.

charlag commented:
@wchristian cool, thanks for helping.  I didn't figure out how to get error message from the server response because response to the failed login is an HTML page as far as I understand. So, what do you show to user? I am not sure that it's cool to show something like 'OkHttpException' because why should we?

wchristian commented:
For one, thanks for the StringUtils suggestion, i've renamed the function and replaced the code with an appropiate StringUtils method.

wchristian commented:
For the other, i'd need to be able to see what actually happens. For the moment i have no secret/id to pass in, nor an idea how to pass it in, so i simply get a GSON parse error.

Once i've seen a broader range of responses i'd be able to formulate a better idea of what to show the user, but in general i suspect it'll be two things:

1. Server response unparsable, please contact the app maintainer. (possible throw the actual server response at Fabric?)
2. Server returned error: <extracted error here, if necessary with regexes>

wchristian commented:
In general i think it's better to at least throw an exception stringification at the user, so the ones smart enough are able to either report it or google it. The ones who aren't will process both a vague error and an exception the same way, so it makes no difference for them.

charlag commented on app/src/main/java/eu/theinvaded/mastondroid/viewmodel/LoginViewModelContract.java:
```diff
@@ -20,7 +20,7 @@
         String getClientId();
         void setUser(MastodonAccount account);
-        void showLoginError();
+        void showLoginError(Throwable loginError);
```
I think View shouldn't care if error is Throwable or something else, it may be better to just pass string or 2 strings for title and message.

wchristian replied:
I have no particular opinion about that. Care to refactor that and link to your idea of how it should look?

charlag replied:
Yeah, okay, I will. It's one of the principles of MVVM: do not pass actual models to the View (or anything that View shouldn't care about). Better stick to the primitives when working with the View layer so view cannot screw up anything and contract stays flexible.

wchristian replied:
Thanks, i'm kinda used to doing that in web apps, but not too familiar with Java yet, so i'd rather learn from how you do that. :)

wchristian commented:
I saw you merged and released a similar branch. I'll rework this one to fit on top of that.

alin-rautoiu commented:
The login part is undercooked mostly because of this

> Please note that the password-based approach is not recommended especially if you're dealing with other user's accounts and not just your own. Usually you would use the authorization grant approach where you redirect the user to a web page on the original site where they can login and authorize the application and are then redirected back to your application with an access code.

https://github.com/tootsuite/mastodon/blob/master/docs/Using-the-API/Testing-with-cURL.md

I took a poor design decision at the beginning. Since the server itself seems to expect this approach, I think the login should be done inside a webview, not trough the app. On the other hand, since it kinda works as it is and some improvements are trivial I considered the move not quite priority.

wchristian commented:
As long as you're happy with taking improvements to it i'm happy to cook it further until it gets replaced at some point.

wchristian committed:
allow users to screenshot login errors by using layout instead of toast

wchristian commented:
Ok, pushed a new version of the branch that builds on the work of the latest merge and brings it up to a similar ux state as the other branch, which i retained in https://github.com/wchristian/mastodroid/tree/old_login_ux

wchristian commented:
Worked out the issue with the /oauth/token path: https://github.com/tootsuite/mastodon/pull/550

alin-rautoiu commented on app/src/main/java/eu/theinvaded/mastondroid/viewmodel/LoginViewModel.java:
```diff
@@ -128,13 +107,23 @@ public void call(Token token) {
                                 isSignIn.set(!isSignIn.get());
-
-                                viewModel.showLoginError();
+                                viewModel.setError("login_process", "login_failed_message", throwable.toString());
```
This looks a bit scary to me. If the resource name is changed in the layout for whatever reason, and it's not reflected here it will crash at runtime.

wchristian committed:
show users concise message for most common login error, details on others

wchristian commented:
@alin-rautoiu I'll add some code to report if that went awry.

Also just added a commit to catch the most common failure response to login and show a useful error, as well as a detailed error if it's not that response.

alin-rautoiu commented on app/src/main/res/layout/activity_login.xml:
```diff
@@ -42,12 +42,60 @@
+                    android:layout_below="@id/login_process_layout"
+                    android:layout_centerInParent="true"
+                    android:textColor="@color/colorPrimaryDark"
+                    android:visibility="@{VIEW.VISIBLE}">
```
I don't think these should be displayed. At some the app should retrieve these automatically. I don't really see why the user should input them by hand.

wchristian replied:
I'm only using them to get a common interface to setErrorEnabled and setError, made them invisibly by having them be zero height. If you have better suggestions, or better styling recommendations, let me know.

Primarily i think using toasts in that area for reporting errors is not great as it precludes users from making screenshots if they're not fast enough.

wchristian commented:
Added a commit that catches errors when trying to report errors on non-existant layouts or error ids, by redirecting them to a hard-coded target, and reporting the error id itself if it couldn't be found in the strings.

wchristian commented:
Fixed up the styling of the error reporting elements by removing the duplicates and removing the unused AppCompatEditText element.

wchristian commented:
Small note: Due to the high number of fix-up commits in here, i intend to refactor the branch a bunch in case you actually decide to merge it, so if you do, please give me a chance for that first.

charlag commented on app/src/main/java/eu/theinvaded/mastondroid/viewmodel/LoginViewModel.java:
```diff
@@ -78,21 +78,8 @@ public void signIn() {
-        } else {
-            viewModel.clearNoPasswordError();
-        }
+        if(!isNonBlankInput("username", viewModel.getUsername())) return;
```
I see what you want to did here but now ViewModel has to know about views IDs and it's not good. Plus, string typing, you know. Maybe there should be static field names in Contract or in somewhere else

charlag commented:
Sorry, I am completely lost here. @wchristian you want to refactor it a little bit more before merging?

wchristian commented:
@charlag can do!

wchristian committed:
experimental conversion of strong ids to static field ids

wchristian commented:
@charlag Refactored the branch to remove "oops" commits and design direction changes, made the intent of commits more clear.

static field names sounds like a good idea, and while i've seen that used in other codes, i'm not too familiar with them or how they're used. I added on a commit experimenting with those. Please let me know how that looks. (I dropped the strings themselves into the class, as reading around putting string constants into the interface seems discouraged.)

charlag commented:
@wchristian thanks but there's a problem with this "experimental id's" - you're breaking contract by referencing activity from ViewModel directly.
I think we should finally do it somehow and start to move to WebView-based auth because of #21

wchristian commented:
> referencing activity from ViewModel directly

What exactly does this mean?

> move to WebView

Sure, but tbh i consider it a mistake to leave the app in a bad ux state until you get that done, unless you think you can do that in 2 days or so.

charlag commented:
@wchristian well, we try to keep components loosely coupled, right? To keep stuff organized properly we've made a contract. ViewModel and View implementations don't know about each other, they only know about the contract.
You're importing fields names in ViewModel from the View and this way you make them tightly coupled.
Geez, I really wish I could just fix it somehow and no, I don't think I can implement new login flow in 2 days. I started to work on the media uploading some time ago but didn't move forward much.

wchristian commented:
> well, we try to keep components loosely coupled, right? To keep stuff organized properly we've made a contract. ViewModel and View implementations don't know about each other, they only know about the contract.
> You're importing fields names in ViewModel from the View and this way you make them tightly coupled.

Alright, that gives me a better idea about your reservations. I have two questions to help me better understand it:

Are you more bothered by the constants not appearing in the contract, or it being constants at all?
Does the string value of the constants being tied to the class bother you and you'd prefer them to be per instance?

> I don't think I can implement new login flow in 2 days

In that case, i think everyone's better served if you focus on the things where you can do good now and help me bang this into good shape and we get that out ASAP, even if it gets ripped out later on.

I also saw the mention of the 2fa stuff, and am willing to experiment with that a bit to see if anything can be done or if i should at least add a note or maybe catch it. Probably in a separate branch though.

charlag commented:
@wchristian I pulled constants to Contract and merged it finally. Sorry that it took *so* long

alin-rautoiu commented:
I pushed a new branch with the new login method. There's a bit more work to be done there (a field for selecting the mastodon instance and removing some hard coded strings). I'll be done with those later today or tomorrow. Until then, could you pull it and test it, to see if there are any problems?

wchristian commented:
@charlag No worries. I learned stuff from it and that's good. :)

@alin-rautoiu Today's not a high energy day for me, so i'll be having a look tomorrow.
